File: leetcode/0001-1000/0643-maximum-average-subarray-i/MaximumAverageSubarrayI.ts
function findMaxAverage(numbers: number[], k: number): number {
  const { max } = numbers.reduce(
    (accumulated, currentValue, currentIndex) => {
      let { max, sum } = accumulated;

      sum += currentValue;

      if (currentIndex >= k) {
        sum -= numbers[currentIndex - k];
      }

      if (currentIndex >= k - 1) {
        max = Math.max(max, sum);
      }

      return { sum, max };
    },
    {
      sum: 0,
      max: Number.NEGATIVE_INFINITY
    }
  )

  return max / k;
};

console.log(findMaxAverage([1, 12, -5, -6, 50, 3], 4))

File: leetcode/0001-1000/0601-0700/0643-maximum-average-subarray-i/MaximumAverageSubarrayI.java
public class MaximumAverageSubarrayI {
  public static void main(String[] args) {
    int[] numbers = new int[] {
        1, 12, -5, -6, 50, 3
    };

    int k = 4;

    System.out.println(new Solution().findMaxAverage(numbers, k));

  }
}

class Solution {
  public double findMaxAverage(int[] numbers, int k) {
    double sum = .0;
    double max = -Double.MAX_VALUE;

    for (int i = 0; i < numbers.length; i++) {
      sum += numbers[i];

      if (i >= k) {
        sum -= numbers[i - k];
      }

      if (i >= k - 1) {
        max = Math.max(max, sum / k);
      }
    }

    return max;
  }
}
